stdlib/domain.ml

```diff
@@ -112,3 +112,40 @@ let get_id { domain; _ } = domain
 
 let self () = Raw.self ()
 
+module DLS = struct
+
+  type 'a key = int ref
+
+  type entry = {key: int ref; slot: Obj.t ref}
+
+  external get_dls_list : unit -> entry list
+    = "caml_domain_dls_get"
+
+  external set_dls_list : entry list  -> unit
+    = "caml_domain_dls_set"
+
+  let new_key () = ref 0
+
+  let set k x =
+    let cs = Obj.repr x in
+    let old = get_dls_list () in
+    let rec add_or_update_entry k v l =
```

Comment:
Can we make this function return `entry option`? The idea is that if the key was already bound, then we'd have updated the value in-place. Hence, no need to invoke `set_dls_list`. Otherwise, a new entry `e` is added to the head. So the function can just return `Some e`. Then, you can change the last line of `set` function to:

```
match add_or_update_entry k cs old with
| None -> ()
| Some e -> set_dls_list (e::old)
```